task/mgmt-gateway/src/update/host_flash.rs

```diff
@@ -0,0 +1,315 @@
+    UpdateStatus,
+};
+
+userlib::task_slot!(HOST_FLASH, hf);
+
+pub(crate) struct HostFlashUpdate {
+    task: HostFlash,
+    current: Option<CurrentUpdate<State>>,
+}
+
+impl HostFlashUpdate {
+    pub(crate) fn new() -> Self {
+        Self {
+            task: HostFlash::from(HOST_FLASH.get_task_id()),
+            current: None,
+        }
+    }
+}
+
+impl ComponentUpdater for HostFlashUpdate {
+    fn block_size(&self) -> usize {
+        static_assertions::const_assert!(
+            PAGE_SIZE_BYTES <= BufferMutex::MAX_CAPACITY
+        );
+        PAGE_SIZE_BYTES
+    }
+
+    fn prepare(
+        &mut self,
+        buffer: &'static BufferMutex,
+        update: ComponentUpdatePrepare,
+    ) -> Result<(), ResponseError> {
+        // Do we have an update already in progress?
+        match self.current.as_ref().map(CurrentUpdate::state) {
+            Some(State::ErasingSectors { .. })
+            | Some(State::AcceptingData { .. }) => {
+                return Err(ResponseError::UpdateInProgress(self.status()));
+            }
+            Some(State::Complete)
+            | Some(State::Aborted)
+            | Some(State::Failed(_))
+            | None => {
+                // All of these states are "done", and we're fine to start a new
+                // update.
+            }
+        }
+
+        // Can we lock the update buffer?
+        let buffer = buffer
+            .borrow(SpComponent::HOST_CPU_BOOT_FLASH, self.block_size())
+            .map_err(|component| {
+                ResponseError::OtherComponentUpdateInProgress(component)
+            })?;
+
+        // Which slot are we updating?
+        let slot = match update.slot {
+            0 => HfDevSelect::Flash0,
+            1 => HfDevSelect::Flash1,
+            _ => return Err(ResponseError::InvalidSlotForComponent),
+        };
+
+        // Do we have control of the host flash?
+        match self
+            .task
+            .get_mux()
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?
+        {
+            HfMuxState::SP => (),
+            HfMuxState::HostCPU => return Err(ResponseError::UpdateSlotBusy),
+        }
+
+        // Swap to the chosen slot.
+        self.task
+            .set_dev(slot)
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+
+        // What is the total capacity of the device?
+        let capacity = self
+            .task
+            .capacity()
+            .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+
+        // How many total sectors do we need to erase? For gimlet, we know that
+        // capacity is an exact multiple of the sector size, which is probably
+        // a safe assumption for future parts as well. We'll assert here in case
+        // that ever becomes untrue, and we can update our math.
+        assert!(capacity % SECTOR_SIZE_BYTES == 0);
+        let num_sectors = (capacity / SECTOR_SIZE_BYTES) as u32;
```

Comment:
Vague question: are we going to support differential writes to host flash (i.e. hashing sectors and only overwriting the ones that changed)?  Or is the network speed so much faster than SWD that it doesn't matter?

Reply:
I was not planning to support differential writes because of the complexity that would add to update delivery. If update performance becomes a concern we can revisit, but I think it's probably fast enough.